modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -4629,3 +4637,166 @@ def from_dataframe(cls, df: "ProtocolDataframe") -> "PandasDataframe":
+
+        parts_len = len(self._partitions)
+        parts = None
+        lengths = None
+        is_trivial_idx = None
+
+        def copartition(df, fill_value):
+            nonlocal is_trivial_idx
+            copartition = False
+            if is_trivial_idx is None:
+                is_trivial_idx = is_trivial_index(self.index)
+            if (
+                is_trivial_idx != is_trivial_index(df.index)
+                or (not is_trivial_idx and any(self.index != df.index))
+                or (parts_len < len(df._partitions))
+            ):
+                copartition = True
+            else:
+                nonlocal parts, lengths
+                if parts is None:
+                    parts = [p[0] for p in self._partitions]
+                    lengths = self._get_lengths(parts, Axis.ROW_WISE)
+
+                df_parts = [p[0] for p in df._partitions]
+                df_part_lengths = df._get_lengths(df_parts, Axis.ROW_WISE)
+                if any(lengths[i] != df_part_lengths[i] for i in range(len(df_parts))):
+                    copartition = True
+            if copartition:
+                _, list_of_right_parts, joined_index, row_lengths = self._copartition(
+                    Axis.ROW_WISE.value,
+                    df,
+                    how="left",
+                    sort=False,
+                    fill_value=fill_value,
+                )
+                df = self.__constructor__(
+                    list_of_right_parts[0],
+                    joined_index,
+                    df.columns,
+                    row_lengths,
+                    df.column_widths,
+                )
+            return df
+
+        # For Dask the callebles are wrapped for each partition in the map_data() function.
+        # If the same callable is wrapped only once for all partitions, CancelledError is raised.
+        wrap_callable = Engine.get() != "Dask"
+        use_map = wrap_callable
+        new_caselist = []
+        for condition, replacement in caselist:
+            if callable(condition):
+                if wrap_callable:
+                    condition = wrapper_put(condition)
+            else:
+                use_map = False
+                if isinstance(condition, cls):
+                    condition = copartition(condition, True)
+            if callable(replacement):
+                if wrap_callable:
+                    replacement = wrapper_put(replacement)
+            elif use_map and is_list_like(replacement):
+                use_map = False
+                if isinstance(replacement, cls):
+                    replacement = copartition(replacement, None)
+            new_caselist.append((condition, replacement))
+
+        # If all the conditions are callable and the replacements are either
+        # callable or scalar, use map().
+        if use_map:
+            return self.map(func=remote_fn, func_args=[new_caselist], lazy=True)
```

Comment:
Why is it beneficial to use lazy=True? Are there any measurements?

Reply:
Answered here - https://github.com/modin-project/modin/pull/6972/#discussion_r1514478027